src/utils/schedule.rs

```diff
@@ -0,0 +1,277 @@
+    /// Using `rrule` library is not sufficient to capture duration of the event so we need to parse it
+    /// manually ane remove it from the string before letting rrule parse the rest
+    /// Duration has to be the last part of the RRULE_SET header after DTSTART e.g.
+    ///   "DTSTART:20221020T180000Z;DURATION:PT1H" not "DURATION:PT1H;DTSTART:20221020T180000Z"
+    /// Duration is in ISO8601 format (`iso8601_duration` crate)
+    fn from_str(calendar_str: &str) -> Result<Self, Self::Err> {
+        let rrule_sets: Vec<&str> = calendar_str
+            .split("DTSTART:")
+            .filter(|s| !s.is_empty())
+            .collect();
+        let mut recurrent_events: Vec<RecurrentEvent> = Vec::new();
+        for rrule_set_str in rrule_sets {
+            let rrules_with_header: Vec<&str> = rrule_set_str
+                .split('\n')
+                .filter(|s| !s.is_empty())
+                .collect();
+            if rrules_with_header.len() < 2 {
+                return Err(());
+            }
+            let header = rrules_with_header[0];
+            let rrules = &rrules_with_header[1..];
+            let header_parts: Vec<&str> = header
+                .split(";DURATION:")
+                .filter(|s| !s.is_empty())
+                .collect();
+            if header_parts.len() != 2 {
+                return Err(());
+            }
+            let dtstart = header_parts[0];
+            let duration = header_parts[1];
+            let str = "DTSTART:".to_owned() + dtstart + "\n" + rrules.join("\n").as_str();
+            let rrule_set = RRuleSet::from_str(&str).unwrap();
+            recurrent_events.push(RecurrentEvent {
+                rrule_set,
+                duration: duration.to_string(),
+            });
+        }
+        Ok(Calendar {
+            events: recurrent_events,
+        })
+    }
+}
+
+impl Display for Calendar {
+    /// Formats `Calendar` into multiline string which can be stored in the database
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let err_msg = String::from("Error writing to string");
+        for event in &self.events {
+            writeln!(
+                f,
+                "DTSTART:{};DURATION:{}",
+                datetime_to_ical_format(event.rrule_set.get_dt_start()),
+                &event.duration
+            )
+            .expect(&err_msg);
+            for rrule in event.rrule_set.get_rrule() {
+                writeln!(f, "RRULE:{}", rrule).expect(&err_msg);
+            }
+            for rdate in event.rrule_set.get_rdate() {
+                writeln!(f, "RDATE:{}", datetime_to_ical_format(rdate)).expect(&err_msg);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Calendar {
+    /// Wrapper implementation of rrule library's `all` method which also considers duration of the event
+    pub fn is_available_between(&self, start_time: DateTime<Tz>, end_time: DateTime<Tz>) -> bool {
```

Comment:
Need more documentation.

Perhaps give an example in the doc that if an event ends at 8, and you pass in a start_time at 8, then this function would return true.

Reply:
I added more documentation